src/session.rs

```diff
@@ -0,0 +1,202 @@
+                    root: root_path.clone(),
+                };
+
+                debug!("Routing editor request to {:?}", route);
+
+                if controllers.contains_key(&route) {
+                    let controller = controllers.get(&route).unwrap();
+                    controller.sender.send(request);
+                } else {
+                    // because Kakoune triggers BufClose after KakEnd
+                    // we don't want textDocument/didClose to spawn new controller
+                    if request.method == notification::DidCloseTextDocument::METHOD {
+                        continue 'event_loop;
+                    }
+                    let controller = spawn_controller(
+                        config.clone(),
+                        route.clone(),
+                        request,
+                        editor_tx.clone(),
+                    );
+                    controllers.insert(route, controller);
+                }
+            }
+        }
+    }
+    stop_session(&mut controllers);
+}
+
+/// Reap controllers associated with editor session.
+fn exit_editor_session(controllers: &mut Controllers, request: &EditorRequest) {
+    info!(
+        "Editor session `{}` closed, shutting down associated language servers",
+        request.meta.session
+    );
+    // have to clone keys as controllers are mutated inside loop
+    for route in controllers.keys().cloned().collect::<Vec<_>>() {
+        if route.session == request.meta.session {
+            // should be safe to unwrap because we are iterating controllers' keys
+            let controller = controllers.remove(&route).unwrap();
+            info!("Exit {} in project {}", route.language, route.root);
+            // to notify kak-lsp about editor session end we use the same `exit` notification as
+            // used in LSP spec to notify language server to exit, thus we can just clone request
+            // and pass it along
+            controller.sender.send(request.clone());
+        }
+    }
+}
+
+/// Shut down all language servers and exit.
+fn stop_session(controllers: &mut Controllers) {
+    let request = EditorRequest {
+        meta: EditorMeta {
+            session: "".to_string(),
+            buffile: "".to_string(),
+            client: None,
+            version: 0,
+        },
+        method: notification::Exit::METHOD.to_string(),
+        params: toml::Value::Table(toml::value::Table::default()),
+    };
+    info!("Shutting down language servers and exiting");
+    for k in controllers.keys().cloned().collect::<Vec<_>>() {
+        // should be safe to unwrap because we are iterating controllers' keys
+        let controller = controllers.remove(&k).unwrap();
+        controller.sender.send(request.clone());
+        info!("Exit {} in project {}", k.language, k.root);
+    }
+    stderr().flush().unwrap();
+    stdout().flush().unwrap();
+    thread::sleep(Duration::from_secs(1));
+    process::exit(0);
```

Comment:
This is a dangerous pattern, calling `process::exit` won't unwind the stack or call destructors, and that is usually desirable. The safe way to use `process::exit` is as a sole call in `main`:

```rust
fn main() {
    let code = main_inner();
    ::std::process::exit(code);
}

fn main_inner() -> i32 {
     ...
}
```

Similarly, not joining a thread can also lead to some info loss. I think the current code has a "lost thread" problem when communicating with language server. We don't wait for this thread:

https://github.com/ul/kak-lsp/blob/9b36a30d595d411059546e4a019718ae1120fec7/src/language_server_transport.rs#L67-L73

That means that, during shutdown, we may kill all controllers and shut down the process *before* these writer threads have a change to write the `exit` notification for proper LSP shutdown. I try to use [structured concurency](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/) pattern in similar cases: if I spawn a thread, I make sure that it joined on all paths by the parent thread.

Reply:
Good catch! I already wanted to eliminate `exit` use, but was unaware of language server writer thread problem. I'll fix it soon.